src/Fmt_ast.ml

```diff
@@ -562,14 +562,24 @@ let sequence_blank_line c xe1 xe2 =
       b - a - commented_lines > 1
   | `Compact -> false
 
+let fits c x ~f =
+  let fmted = Cmts.preserve f x in
+  3 * String.length fmted < c.conf.margin
+  &&
+  match String.rindex fmted '\n' with
+  | Some i -> i = 0 (* if the first character is '\n' we ignore it *)
+  | None -> true
+
 let rec fmt_extension c ctx key (ext, pld) =
   match (pld, ctx) with
   | ( PStr [({pstr_desc= Pstr_value _ | Pstr_type _; _} as si)]
     , (Pld _ | Str _ | Top) ) ->
       fmt_structure_item c ~last:true ~ext (sub_str ~ctx si)
   | PSig [({psig_desc= Psig_type _; _} as si)], (Pld _ | Sig _ | Top) ->
       fmt_signature_item c ~ext (sub_sig ~ctx si)
-  | _ -> fmt_attribute_or_extension c key Fn.id (ext, pld)
+  | _ ->
+      let fits = fits c pld ~f:(fmt_payload c (Pld pld)) in
```

Comment:
Why is this necessary ? Is this not possible to implement with `fits_breaks` ?

Reply:
No because we would need to open the box before formatting (before using fits_breaks)